cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/proto/ValueFilter.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.report.proto;
+
+import co.cask.cdap.report.util.ReportField;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Represents a filter that checks whether a given value of a field is one of the allowed values and is not one of
+ * the forbidden values.
+ *
+ * @param <T> type of the values
+ */
+public class ValueFilter<T> extends Filter<T> {
+  @Nullable
+  private final List<T> whitelist;
+  @Nullable
+  private final List<T> blacklist;
+
+  public ValueFilter(String fieldName, @Nullable List<T> whitelist, @Nullable List<T> blacklist) {
+    super(fieldName);
+    this.whitelist = whitelist;
+    this.blacklist = blacklist;
+  }
+
+  /**
+   * @return the allowed values of this field, or {@code null} if there's no such limit
+   */
+  @Nullable
+  public List<T> getWhitelist() {
+    return whitelist;
+  }
+
+  /**
+   * @return the disallowed values of this field, or {@code null} if there's no such limit
+   */
+  @Nullable
+  public List<T> getBlacklist() {
+    return blacklist;
+  }
+
+  @Override
+  public List<String> getErrors() {
+    ArrayList<String> errors = new ArrayList<>();
+    errors.addAll(super.getErrors());
+    errors.addAll(getFilterTypeErrors(ReportField.FilterType.VALUE));
+    if (whitelist == null && blacklist == null) {
+      errors.add(String.format("'whitelist' and 'blacklist' cannot both be null in filter '%s'", getFieldName()));
+    } else if (whitelist != null && blacklist != null) {
+      Set<T> duplicates = new HashSet(whitelist);
+      duplicates.retainAll(blacklist);
+      if (!duplicates.isEmpty()) {
+        errors.add(String.format("'whitelist' and 'blacklist' should not contain duplicated elements in filter '%s'",
+                                 getFieldName()));
+      }
+    }
+    return errors;
+  }
+
+  @Override
+  public boolean apply(T value) {
+    return (whitelist == null || whitelist.isEmpty() || whitelist.contains(value))
+      && (blacklist == null || blacklist.isEmpty() || !blacklist.contains(value));
+  }
+
+  @Override
+  public String toString() {
+    return "ValueFilter{" +
+      "fieldName=" + getFieldName() +
+      ", whitelist=" + whitelist +
```

Comment:
you need to use joiner to join the list contents right

Reply:
toString method of List automatically joins the list with ", " and enclose the list in []